### DISCUSSION TITLE
The issue to invoke the solution after the model is solved

### DISCUSSION THREAD

**Discussion by abb-omidi** [PyVRP/PyVRP#654] [Q&A] [answered]
Dear support team,

I hope you are doing well. I am trying to solve an instance of the problem and I am having a dificalty to invoke the solution. Let me explain more:

After initializing the input data and solving the model the output seems to be correct. The output is something like this:
```
print(best_solution)

Route #1: 9 2 5 
Route #2: 7 6 1 
Route #3: 4 10 3 
Route #4: 8 
```
Now, I would like to get this solution as a nested list based on the vehicles or something like that as a dictionary. 
* Would you please, is there any way to take the solution as a list or dictionary something like this: [[9,2,5],[7,6,1],[4,10,3],[8]]

Also, as a follow up questions, 
* Does the attribute `release_time` inside `clients` method use a single array of released date of each order? (It seems it can)
* What is the difference between defining the `release_time` as an array like this `[10,20]` or as a `TIME_WINDOWS` something like this `(10,10),(20,20)`? (obviously the solutions are diffrent, at least in my case and I would like to know why?)

All the best
Abbas

**Comment by N-Wouda** [comment, marked as answer]
Hi @abb-omidi 

>Would you please, is there any way to take the solution as a list or dictionary something like this: [[9,2,5],[7,6,1],[4,10,3],[8]]

The easiest is probably
```python
visits = [route.visits() for route in best_solution.routes()]
```

>Does the attribute release_time inside clients method use a single array of released date of each order? (It seems it can)

`release_time` is an integer that specifies the earliest time such a client order may leave the depot, if it's put on a route. It thus impacts route's start times. I do not understand the question about arrays.

>What is the difference between defining the release_time as an array like this [10,20] or as a TIME_WINDOWS something like this (10,10),(20,20)? (obviously the solutions are diffrent, at least in my case and I would like to know why?)

Release times are not time windows. Release times say something about when a client order may leave the depot. The time windows say something about when we may arrive at the client location.

**Comment by abb-omidi** [comment]
Dear @N-Wouda,

Many thanks for your comments. I tried with the `best_solution`, but it actually needs to use `routes` method inside that! 
For `Release times` and based on my previous comments and your answer it is: `Can be modeled as earliest time windows in VRPs.`. 
Would you say please, is it a new thing in the `pyvrp-0.9.1`, or maybe I am missing something?  

All the best

**Reply by N-Wouda** [reply]
> For `Release times` and based on my previous comments and your answer it is: `Can be modeled as earliest time windows in VRPs.`.

Release times are *not* earliest time windows. They are completely different concepts.  

> Would you say please, is it a new thing in the `pyvrp-0.9.1`, or maybe I am missing something?

Release times have been in PyVRP since v0.4.0 or so; they are not new.

**Comment by abb-omidi** [comment]
@N-Wouda,

Many thanks for the clarification. 
Might the precedence relation be implemented in the new version `pyvrp-0.9.1` or may not?

Best

**Reply by N-Wouda** [reply]
Precedence relations are not planned for the upcoming release.
